### PR TITLE
workplace http request handler implementation

### PR DESCRIPTION
**Pull request by ndmytro**

### CODE REVIEW

**Comment by Fedorof** [comment]
s there a decision to create another app named `api`?

File: SSP/api/views.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
+from django.shortcuts import HttpResponse
+from carcassonne.models import Workplace
+import json
+
+def workplace(request):
+    """
+    Write to or get from database information concerning workplaces depending on request method.
```
Comment (Fedorof) [inline SSP/api/views.py]: PEP257

File: SSP/ssp_project/settings.py [hunk 2]
```diff
@@ -23,7 +23,7 @@
 SECRET_KEY = 'a8_f#x42#vyy%rs)doj!i@1j3no=bp#*(w$7y5sgt0v_2%wf--'
 
 # SECURITY WARNING: don't run with debug turned on in production!
-DEBUG = False
+DEBUG = True
```
Comment (Fedorof) [inline SSP/ssp_project/settings.py]: No, it is set in dev settings

File: SSP/api/views.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        number = json_request.get('number')
+        coordinates = json_request.get('coordinates')
+        status = json_request.get('status')
+        if not (room and number and coordinates and status):
+            return HttpResponse(status=400)
+        workplace = Workplace(room=room, number=number, coordinates=coordinates, status=status)
+        workplace.save()
+        return HttpResponse(status=201)
+
+    if request.method == 'GET':
```
Comment (Fedorof) [inline SSP/api/views.py]: from django.views.decorators.http import require_http_methods

    @require_http_methods(['POST'])
    def workplace(request):
        ...

File: SSP/api/views.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+    :return: HTTP status 201 is returned and the JSON is utilized if all the fields in required JSON are present.
+    HTTP status 400 is returned in case of any errors.
+    """
+    if request.method == 'POST':
+        json_request = json.loads(request.body.decode('utf-8'))
+        room = json_request.get('room')
+        number = json_request.get('number')
+        coordinates = json_request.get('coordinates')
+        status = json_request.get('status')
```
Comment (Fedorof) [inline SSP/api/views.py]: Could you please add blank lines to separate logical blocks of the code?

File: SSP/api/views.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        if not (room and number and coordinates and status):
+            return HttpResponse(status=400)
```
Comment (Fedorof) [inline SSP/api/views.py]: Perhaps return `JsonResponce` with `status_code` set to 422 and contents like this:

    {'error':
        {
         'code': 422,
         'type': 'Unprocessable Entity',
         'message': 'Missing one or more of the required parameters: room, number, coordinates, status'
        }
}

**Comment by Fedorof** [comment]
And tests, please!

File: README.md [hunk 6]
```diff
@@ -1,7 +1,10 @@
 #To open plan editor run the server and open http://127.0.0.1:8000/edit_plan/
 1. Clone this project
-2. vagrant up
+2. vagrant up --provision (migrate command will invoke fixture initial_data.json)
```
Comment (Fedorof) [inline README.md]: Perhaps applying fixtures should not go to provision

File: SSP/api/tests.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
+from django.test import TestCase
```
Comment (Fedorof) [inline SSP/api/tests.py]: I would create a package `tests` with sub-files like specified in `unittest` documentation. Test methods should follow the pattern `test_*` as well.

File: SSP/api/tests.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        self.client = Client()
+        office = Office.objects.create(
+            name="Kiev1",
+            address="Some Address"
+        )
+        floor = Floor.objects.create(number=2, office=office)
+        Room.objects.create(number=225, floor=floor)
+
+    def test1(self):
+        """Test POST request in case when all the required data is provided and
```
Comment (Fedorof) [inline SSP/api/tests.py]: PEP 257

File: SSP/api/tests.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        """
+        json_data = {
+            "office": "Kiev1",
+            "floor": "2",
+            "room": "225",
+            "number": "5",
+            "coordinates":
+            "[[4.375,15.75],[4.375,18],[7.78125,18],[7.78125,15.75],[4.375,15.75]]",
+            "status": "True"
+                    }
```
Comment (Fedorof) [inline SSP/api/tests.py]: I would dedent this line

File: SSP/api/tests.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+            "floor": "2",
+            "number": "5",
+            "coordinates":
+            "[[4.375,15.75],[4.375,18],[7.78125,18],[7.78125,15.75],[4.375,15.75]]",
+            "status": "True"
+                    }
+
+        response = self.client.post('/api/workplaces/', json.dumps(json_data),
+                                    content_type='application/json')
+        self.assertEqual(response.status_code, 422)
```
Comment (Fedorof) [inline SSP/api/tests.py]: Should not the error contain additional information? See here for example: https://github.com/ITsvetkoFF/Kv-011/pull/61/files#diff-8aac644b03d969355461e4aa23350fb0R23

File: SSP/api/tests.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+            "room": "225",
+            "number": "5",
+            "coordinates":
+            "[[4.375,15.75],[4.375,18],[7.78125,18],[7.78125,15.75],[4.375,15.75]]",
+            "status": "True"
+                    }
+
+        response = self.client.post('/api/workplaces/', json.dumps(json_data),
+                                    content_type='application/json')
+        self.assertEqual(response.status_code, 406)
```
Comment (Fedorof) [inline SSP/api/tests.py]: The same as above

File: SSP/api/tests.py [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+            "room": "225",
+            "number": "5",
+            "coordinates":
+            "[[4.375,15.75],[4.375,18],[7.78125,18],[7.78125,15.75],[4.375,15.75]]",
+            "status": "True"
+                    }
+
+        response = self.client.post('/api/workplaces/', json.dumps(json_data),
+                                    content_type='application/json')
+        self.assertEqual(response.status_code, 201)
```
Comment (Fedorof) [inline SSP/api/tests.py]: How do you check if the changes took place?

File: SSP/api/tests.py [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+            "[[4.375,15.75],[4.375,18],[7.78125,18],[7.78125,15.75],[4.375,15.75]]",
+            "status": "True"
+                    }
+
+        response = self.client.post('/api/workplaces/', json.dumps(json_data),
+                                    content_type='application/json')
+        self.assertEqual(response.status_code, 406)
+
+
+
```
Comment (Fedorof) [inline SSP/api/tests.py]: 3 empty lines. Remove them, please

File: SSP/api/tests.py [hunk 14]
```diff
@@ -0,0 +1,72 @@
+from django.test import TestCase
+from django.test import Client
+from carcassonne.models import Office, Floor, Room, Workplace
+import json
```
Comment (Fedorof) [inline SSP/api/tests.py]: PEP8 for imports:

    import system modules
    <blank line>
    import third-party modules
    <blank line>
    import our own modules

File: SSP/api/views.py [hunk 15]
```diff
@@ -0,0 +1,104 @@
+from django.shortcuts import HttpResponse
+from carcassonne.models import Workplace, Room, Floor, Office
+from django.views.decorators.csrf import csrf_exempt
+import json
+from django.views.decorators.http import require_http_methods
+from django.http import JsonResponse
```
Comment (Fedorof) [inline SSP/api/views.py]: PEP8 for imports (see above)

File: SSP/api/views.py [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+        "[[4.375,15.75],[4.375,18],[7.78125,18],[7.78125,15.75],[4.375,15.75]]",
+    "status": "True"
+    }
+
+    :return: HTTP status 201 is returned and the JSON is utilized if all
+    the fields in required JSON are present.
+    In case of any errors a JSON with further information is returned.
+    """
+
+    if not request.body:
```
Comment (Fedorof) [inline SSP/api/views.py]: Do you really need this check? Seems like it's to weak. Perhaps `400 Bad Request` if you can't parse JSON

File: SSP/api/views.py [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+            'Missing one or more of the required parameters: '
+            'office, floor, room, number, coordinates, status')
+
+
+    office = Office.objects.filter(name=office_name).first()
+    floor = Floor.objects.filter(number=floor).first()
+
+    if not(office and floor):
+        return error_resp(
+            406,
```
Comment (Fedorof) [inline SSP/api/views.py]: 406 is for not acceptable format. 422 or 400 goes better.

File: SSP/api/views.py [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+            422,
+            'Unprocessable Entity',
+            'Missing one or more of the required parameters: '
+            'office, floor, room, number, coordinates, status')
+
+
+    office = Office.objects.filter(name=office_name).first()
+    floor = Floor.objects.filter(number=floor).first()
+
+    if not(office and floor):
```
Comment (Fedorof) [inline SSP/api/views.py]: `and` or `or`? Do you have tests to check for missing `office` and `floor` separately?

File: SSP/api/views.py [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+        return error_resp(
+            406,
+            "Not Acceptable",
+            'One of the parameters (office, floor) does not correspond to'
+            ' any existing database entry')
```
Comment (Fedorof) [inline SSP/api/views.py]: The leading whitespace should go to the previous line.

File: SSP/api/views.py [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+            and coordinates and status):
+        return error_resp(
+            422,
+            'Unprocessable Entity',
+            'Missing one or more of the required parameters: '
+            'office, floor, room, number, coordinates, status')
+
+
+    office = Office.objects.filter(name=office_name).first()
+    floor = Floor.objects.filter(number=floor).first()
```
Comment (Fedorof) [inline SSP/api/views.py]: Should not you have to get floor for the current office? Do you have tests for the case when there is more than one office?

File: SSP/api/views.py [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+    if not (office_name and floor and room_number and workplace_number
+            and coordinates and status):
+        return error_resp(
+            422,
+            'Unprocessable Entity',
+            'Missing one or more of the required parameters: '
+            'office, floor, room, number, coordinates, status')
+
+
+    office = Office.objects.filter(name=office_name).first()
```
Comment (Fedorof) [inline SSP/api/views.py]: Should not it be `get`?

File: SSP/api/views.py [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+            406,
+            "Not Acceptable",
+            'One of the parameters (office, floor) does not correspond to'
+            ' any existing database entry')
+
+    room, _ = Room.objects.get_or_create(number=room_number, floor=floor)
+
+    workplace = Workplace(room=room, number=workplace_number,
+                          coordinates=coordinates, status=status)
+    workplace.save()
```
Comment (Fedorof) [inline SSP/api/views.py]: Would `Workplace.objects.create` be better in this place?

File: SSP/ssp_project/settings.py [hunk 23]
```diff
@@ -42,7 +42,6 @@
     'carcassonne',
     'jsonfield',
     'carcassonne.helpers.wms',
-    'djangobower',
```
Comment (Fedorof) [inline SSP/ssp_project/settings.py]: Is it really should be removed?

File: SSP/ssp_project/urls.py [hunk 24]
```diff
@@ -26,4 +26,5 @@
     url(r'^logout/$', auth_views.logout, name='logout'),
     url(r'^edit_plan/', include('plan_editor.urls')),
     url(r'^plan_viewer/', include('carcassonne.urls')),
-]+ static(settings.STATIC_URL, document_root=settings.STATIC_ROOT)
```
Comment (Fedorof) [inline SSP/ssp_project/urls.py]: Is it really should be removed?

File: SSP/carcassonne/models.py [hunk 25]
```diff
@@ -28,9 +28,10 @@ def __str__(self):
 class Room(models.Model):
     """
     Store data associated with Room entity: coordinates, room number
+
     Related to model: 'carcassonne.Floor'
     """
-    coordinates = JSONField()
+    coordinates = JSONField(blank=True)
```
Comment (ITsvetkoFF) [inline SSP/carcassonne/models.py]: model is changed so we need migration :)

Comment (ITsvetkoFF) [inline SSP/carcassonne/models.py]: And then I think you can squash commits and we can merge this

File: SSP/api/views.py [hunk 26]
```diff
@@ -0,0 +1,106 @@
...
+    coordinates = json_request.get("coordinates")
+    is_occupied = json_request.get("is_occupied")
+
+    if not (office_name and floor and room_number and workplace_number
+            and coordinates and is_occupied):
+        return error_resp(
+            422,
+            'Unprocessable Entity',
+            'Missing one or more of the required parameters: '
+            'office, floor, room, number, coordinates, status')
```
Comment (Iodina) [inline SSP/api/views.py]: There should be is_occupied instead of status now

File: SSP/api/tests/test_workplace_api.py [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+
+from django.test import TestCase
+from django.test import Client
+
+from carcassonne.models import Office, Floor, Room, Workplace
+
+
+class CreateWorkplaceTestCase(TestCase):
+    def setUp(self):
+        """Create instances of Office, Floor and Room in DB - one
```
Comment (Fedorof) [inline SSP/api/tests/test_workplace_api.py]: PEP 257. Here and below

File: SSP/api/tests/test_workplace_api.py [hunk 28]
```diff
@@ -0,0 +1,99 @@
...
+            "number": 5,
+            "coordinates":
+                "[[4.375,15.75],[4.375,18],[7.78125,18],[7.78125,15.75],[4.375,15.75]]",
+            "is_occupied": "True"
+        }
+
+        self.client.post('/api/workplaces/', json.dumps(json_data),
+                                    content_type='application/json')
+
+        self.assertIsInstance(Workplace.objects.get(number=5), Workplace,
```
Comment (Fedorof) [inline SSP/api/tests/test_workplace_api.py]: This check is meaningless. If Workplace does not exist it raises `ObjectDoesNotExist`

File: SSP/api/tests/test_workplace_api.py [hunk 29]
```diff
@@ -0,0 +1,99 @@
...
+            "number": 5,
+            "coordinates":
+                "[[4.375,15.75],[4.375,18],[7.78125,18],[7.78125,15.75],[4.375,15.75]]",
+            "is_occupied": "True"
+        }
+
+        response = self.client.post('/api/workplaces/', json.dumps(json_data),
+                                    content_type='application/json')
+        self.assertEqual(response.status_code, 422,
+                         msg="Test failed! JSON was missing parameters, "
```
Comment (Fedorof) [inline SSP/api/tests/test_workplace_api.py]: It masks the real status code. Either remove the `msg` or provide more meaningful information like status code and message.

Here and below.

File: SSP/api/tests/test_workplace_api.py [hunk 30]
```diff
@@ -0,0 +1,99 @@
...
+            "floor": 2,
+            "room": "225",
+            "number": 5,
+            "coordinates":
+                "[[4.375,15.75],[4.375,18],[7.78125,18],[7.78125,15.75],[4.375,15.75]]",
+            "is_occupied": "True"
+        }
+
+        self.client.post('/api/workplaces/', json.dumps(json_data),
+                                    content_type='application/json')
```
Comment (Fedorof) [inline SSP/api/tests/test_workplace_api.py]: PEP8

Comment (Fedorof) [inline SSP/api/tests/test_workplace_api.py]: Assert that response code is 201?

File: SSP/api/views.py [hunk 31]
```diff
@@ -0,0 +1,106 @@
...
+        {'error':
+            {
+                'code': code,
+                'type': error_type,
+                'message': message,
+            }
+        },
+        status=code
+    )
+
```
Comment (Fedorof) [inline SSP/api/views.py]: Please, add another blank line (PEP8)

File: SSP/api/views.py [hunk 32]
```diff
@@ -0,0 +1,106 @@
...
+            return error_resp(
+                422,
+                'Unprocessable Entity',
+                'One of the parameters (office, floor) does not correspond to'
+                ' any existing database entry')
+
+    room, _ = Room.objects.get_or_create(number=room_number, floor=floor)
+
+    Workplace.objects.create(room=room, number=workplace_number,
+                          coordinates=coordinates, is_occupied=is_occupied)
```
Comment (Fedorof) [inline SSP/api/views.py]: PEP8 - incorrect indent.

**Pull request merged**